### PR TITLE
feat(data-classes): add AttributeValueType to DynamoDBStreamEvent

### PR DESCRIPTION
**Pull request by michaelbrewer**
**Issue #, if available:**

https://github.com/awslabs/aws-lambda-powertools-python/issues/461

## Description of changes:

Changes:
- [x] Add new enum AttributeValueType for the type of AttributeValue
- [x] Add new property `get_type` to return said enum
- [x] Add new property `get_value` which returns converted value
- [x] Update docs with any revisited changes in the type conversion process

Example usage:

```python3
from aws_lambda_powertools.utilities.data_classes import event_source, DynamoDBStreamEvent
from aws_lambda_powertools.utilities.data_classes.dynamo_db_stream_event import AttributeValueType, AttributeValue
from aws_lambda_powertools.utilities.typing import LambdaContext


@event_source(data_class=DynamoDBStreamEvent)
def lambda_handler(event: DynamoDBStreamEvent, context: LambdaContext):
    for record in event.records:
        key: AttributeValue = record.dynamodb.keys["id"]
        if key == AttributeValueType.Number:
            assert key.get_value == key.n_value
            print(key.get_value)
        elif key == AttributeValueType.Map:
            assert key.get_value == key.map_value
            print(key.get_value)
```

Where `AttributeValueType` is:

```python3
class AttributeValueType(Enum):
    Binary = "B"
    BinarySet = "BS"
    Boolean = "BOOL"
    List = "L"
    Map = "M"
    Number = "N"
    NumberSet = "NS"
    Null = "NULL"
    String = "S"
    StringSet = "SS"
```

## How AttributeValue does the raw values

Example Values:

* b_value:
```{"B": "dGhpcyB0ZXh0IGlzIGJhc2U2NC1lbmNvZGVk"}```
* bs_value
```{"BS": ["U3Vubnk=", "UmFpbnk=", "U25vd3k="]}```
* bool_value
```{"BOOL": True}```
* list_value/l_value
```{"L": [ {"S": "Cookies"} , {"S": "Coffee"}, {"N": "3.14159"}]}```
* map_value/m_value
```{"M": {"Name": {"S": "Joe"}, "Age": {"N": "35"}}}```
* n_value
```{"N": "123.45"}```
* ns_value
```{"NS": ["42.2", "-19", "7.5", "3.14"]}```
* null_value
```{"NULL": True}```
* s_value
```{"S": "Hello"}```
* ss_value
```{"SS": ["Giraffe", "Hippo" ,"Zebra"]}```

> NOTE: All return types are `Optional`

Return types

| Property           | Description                                                      | Raw Return Type             |
|--------------------|------------------------------------------------------------------|-----------------------------|
| b_value            | An attribute of type Base64-encoded binary data object           | str                         |
| bs_value           | An attribute of type Array of Base64-encoded binary data objects | List[str]                   |
| bool_value         | An attribute of type Boolean                                     | bool                        |
| list_value l_value | An attribute of type Array of AttributeValue objects             | List["AttributeValue"]      |
| map_value m_value  | An attribute of type String to AttributeValue object map         | Dict[str, "AttributeValue"] |
| n_value            | An attribute of type Number                                      | str                         |
| ns_value           | An attribute of type Number Set                                  | List[str]                   |
| null_value         | An attribute of type Null                                        | bool                  |
| s_value            | An attribute of type String                                      | str                         |
| ss_value           | An attribute of type Array of strings                            | List[str]                   |

## How Dynamodb Type does deserialization

| DynamoDB	|	Python |
|-----------|----------|
| {'NULL': True}	|	None |
| {'BOOL': True/False}	|	True/False |
| {'N': str(value)}	|	Decimal(str(value)) |
| {'S': string}	|	string |
| {'B': bytes}	|	Binary(bytes) |
| {'NS': [str(value)]}	|	set([Decimal(str(value))]) |
| {'SS': [string]}	|	set([string]) |
| {'BS': [bytes]}	|	set([bytes]) |
| {'L': list}	|	list |
| {'M': dict}	|	dict |

**Checklist**

 

* [X] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [X] Update tests
* [X] Update docs
* [X] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by michaelbrewer** [comment]
@gwlester what do you think?

**Comment by gwlester** [comment]
Looks great!

Thanks for the fast work!

**Review by gwlester — APPROVED** [review]

File: aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py [hunk 1]
```diff
@@ -4,12 +4,30 @@
...
+
+
 class AttributeValue(DictWrapper):
     """Represents the data for an attribute
 
     Documentation: https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_AttributeValue.html
     """
 
+    @property
+    def get_type(self) -> AttributeValueType:
```
Comment (jplock) [inline aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:27]: What about a generic “get_value” method that can inspect the type and return the proper value?

Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:27]: @jplock -  But then i would need to type hint as a Union? Coz it is kind of nice having the type hinting

Comment (jplock) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:27]: That’s true. I end up using the boto3.dynamodb.types TypeDeserializer directly anyway

Comment (gwlester) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:27]: I happen to need the **get_type**, but a **get_value** is an obvious idea, should have also suggested that.

Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:27]: Thanks @jplock for pointing me to https://boto3.amazonaws.com/v1/documentation/api/latest/_modules/boto3/dynamodb/types.html

I am thinking of a type hinted version of this.

Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py]: @jplock @gwlester @heitorlessa 

I added "value" implementation, as it is not quite a deserialize. But maybe it should be and do the binary handling to? However that would be a breaking change.

File: aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py [hunk 2]
```diff
@@ -1,15 +1,39 @@
 from enum import Enum
-from typing import Dict, Iterator, List, Optional
+from typing import Any, Dict, Iterator, List, Optional, Union
 
 from aws_lambda_powertools.utilities.data_classes.common import DictWrapper
 
 
+class AttributeValueType(Enum):
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:7]: Could you update the DynamoDB Stream data class doc to include an example usage?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
I'll release 1.17.0 tomorrow and I could get this included too, if we could have an example usage in the docs. Thanks a lot!

**Comment by gwlester** [comment]
Will this be in the next Release?

File: aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py (hunk 1 above)
Comment (heitorlessa) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py]: I'm fine either way. 'get_value' would be more consistent and inline with the operation it does here than 'value'

File: aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py (hunk 2 above)
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:7]: Ok sure.

Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py:7]: @heitorlessa CI process seems to be broken

File: aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py (hunk 1 above)
Comment (michaelbrewer) [reply aws_lambda_powertools/utilities/data_classes/dynamo_db_stream_event.py]: @heitorlessa renamed

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
Thanks @michaelbrewer for the changes! 

I'm gonna look into the CI failures

**Comment by michaelbrewer** [comment]
> Thanks @michaelbrewer for the changes!
> 
> I'm gonna look into the CI failures

Failures seem to be intermittent getting the codecov token

**Comment by heitorlessa** [comment]
Yeah I raised a ticket back then and seemed resolved for a while.

I also noticed concurrent doc builds fail due to git divergence when using
Mike

**Comment by gwlester** [comment]
FYI, added it into the code (Dynamo Stream -> Lambda -> TimeStream -- for audit trail) and it works great!!!

Thanks again!

**Comment by heitorlessa** [comment]
Fantastic to hear!

**Pull request merged**